code/globalincs/pool.h

```diff
@@ -0,0 +1,404 @@
+				return tl::nullopt;
+			}
+			return add_new(n);
+		}
+		return use_free(n);
+	};
+
+	void remove(optional<pool_index> i){
+		if (!i.has_value())
+			return;
+		remove(i.value());
+	};
+	void remove(pool_index i){
+		if (i.null())
+			throw PoolExceptionNullIndex("Attempted to remove from SCP_Pool with a null index");
+		if (i.i() >= storage.size())
+			throw PoolExceptionOutOfRange("Attempted to remove from SCP_Pool with out-of-range index");
+		if (storage[i.i()].generation != i.g() ) 
+			throw PoolExceptionStaleIndex("Attempted to remove from SCP_Pool with stale index");
+		reset_slot(i.i());
+	};
+	bool check(pool_index i){
+		if(i.null())
+			return false;
+		auto * t = &operator[](i);
+		return (t!=nullptr);
+	}
+
+
+
+	//iterator shit.
+	struct iter{
+		using iterator_category = std::input_iterator_tag;
+		using index = int;
+		using storage_t = vector<gEntry<T>>;
+		using storage_p = vector<gEntry<T>>*;
+
+		iter(SCP_Pool<T> *vec_ptr, storage_p storage_ptr){
+			storage = storage_ptr;
+			vec = vec_ptr;
+		};
+		iter(SCP_Pool<T> *vec_ptr, storage_p storage_ptr, int start_pos){
+			storage = storage_ptr;
+			vec = vec_ptr;
+			inner_position = start_pos;
+			if(inner_position < storage->size())
+			{
+				if(storage->at(inner_position).stored == tl::nullopt){
+					(*this)++;
+				}
+				else {
+					full_position =pool_index(inner_position,storage->at(inner_position).generation);
+				}
+			}
+		};
+		bool operator==(const iter& rhs){
+			return inner_position == rhs.inner_position;
+		};
+		bool operator<(const iter& rhs){
+			return inner_position < rhs.inner_position;
+		};
+		iter& operator++(){
+			inner_position++;
+			//we want to skip empty slots
+			while(inner_position < storage->size()
+					&& storage->at(inner_position).stored == tl::nullopt 
+				  ){
+				inner_position++;
+				//the while condition isn't short-circuiting so we're getting invalid accesses.
+				if(inner_position==storage->size()) break;
+			} 
+			if(inner_position < storage->size()){
+			   if(storage->at(inner_position).stored != tl::nullopt){
+				full_position.set(inner_position,storage->at(inner_position).generation);
+			   }
+			   else{
+				 full_position.nullify();
+			   }
+			}
+			else{full_position.nullify();}
+			return *this;
+		};
+		iter& operator++(int i){
+			operator++();
+			return *this;
+		}
+		T* operator*(){
+			//vec->get_pointer(full_position);
+			if(vec->check(full_position))
+				return &vec->operator[](full_position);
+			return nullptr;
+		};
+		T& operator&(){
+//			T t = *vec[full_position];
+			return vec->operator[](full_position);
+		};
+		pool_index position(){return full_position;}
+	private:
+		index inner_position;
+		pool_index full_position;
+		storage_p storage;
+		SCP_Pool<T> *vec;
+	};
+	SCP_Pool<T>::iter begin(){
+		return iter(this,&storage,0);
+		//if (known_empty_) {
+		//statements
+		//}
+	};
+	SCP_Pool<T>::iter end(){
+		return iter(this,&storage,(int) storage.size());
+	}
+};
```

Comment:
Now that we've had so much fun with this iterator, how about we do it again.
On a serious note, having a const_iterator that does basically the same except returning const references is absolutely needed here.
Both are cbegin and cend useful overloads to enable optimizations, having a ``SCP_Pool<T>::const_iterator begin() const;`` overload is cruical to allow foreaching and iterating the pool even when we just have a const reference to it, which tends to be often.